graph/src/components/metrics/stopwatch.rs

```diff
@@ -0,0 +1,169 @@
+use crate::prelude::*;
+use std::collections::HashMap;
+use std::sync::{atomic::AtomicBool, atomic::Ordering, Mutex};
+use std::time::Instant;
+
+/// This is a "section guard", that closes the section on drop.
+pub struct Section {
+    id: String,
+    stopwatch: StopwatchMetrics,
+}
+
+impl Section {
+    /// A more readable `drop`.
+    pub fn end(self) {}
+}
+
+impl Drop for Section {
+    fn drop(&mut self) {
+        self.stopwatch
+            .end_section(std::mem::replace(&mut self.id, String::new()))
+    }
+}
+
+/// Usage example:
+/// ```ignore
+/// // Start counting time for the "main_section".
+/// let _main_section = stopwatch.start_section("main_section");
+/// // do stuff...
+/// // Pause timer for "main_section", start for "child_section".
+/// let chid_section = stopwatch.start_section("child_section");
+/// // do stuff...
+/// // Register time spent in "child_section", implicitly going back to "main_section".
+/// section.end();
+/// // do stuff...
+/// // At the end of the scope `_main_section` is dropped, which is equivalent to calling
+/// // `_main_section.end()`.
+#[derive(Clone)]
+pub struct StopwatchMetrics {
+    disabled: Arc<AtomicBool>,
+    inner: Arc<Mutex<StopwatchInner>>,
+}
+
+impl StopwatchMetrics {
+    pub fn new(
+        logger: Logger,
+        subgraph_id: SubgraphDeploymentId,
+        registry: Arc<dyn MetricsRegistry>,
+    ) -> Self {
+        let mut inner = StopwatchInner {
+            total_counter: *registry
+                .new_counter(
+                    format!("{}_total_secs", subgraph_id),
```

Comment:
I'm not sure `total_secs` is the right term. Maybe `{}_secs_to_synced` or `{}_syncing_secs`? Who knows what we might use `total` for in the future, maybe _all_ time spent on a subgraph?

Reply:
Maybe this is fine actually, as the stopwatch metrics could be used in other places too. But then I'd maybe pass in a `sync` prefix here that allows this to become `{}_sync_total_secs`.

Reply:
Interesting, I hadn't considered the reuse potential. Should all metrics be prefixed with `_sync` or just the one for the totals?

Reply:
Initially, I thought we should prefix all but I think just the totals works. `{subgraph}_total_secs` is just too generic, but `{subgraph}_store_get` (or whatever) makes sense even if it's just capture during syncing for now. What do you think?

Reply:
Agreed, I'll change just the `total_secs` to `sync_total_secs`.